OPN ops TupleConstruct/Unpack and format.

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29664 Add missing operators for PyText model.
* #29656 Add overload name to JIT prim operators.
* **#29635 OPN ops TupleConstruct/Unpack and format.**

TupleConstruct/Unpack as OPN ops.

Differential Revision: [D18499602](https://our.internmc.facebook.com/intern/diff/D18499602)

Commit:
OPN ops TupleConstruct/Unpack and format.

Commit:
Update on "OPN ops TupleConstruct/Unpack and format."

Commit:
Update on "OPN ops TupleConstruct/Unpack and format."

ljk53 commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -26,6 +27,57 @@ void tensorListConstruct(int num_inputs, Stack& stack) {
+     tuple->elements().end());
+}
+
+static const std::regex unsupported_options("\\{(.*?)\\}");
```
nit: we can move this inside format() function if this is only used by it?

ljk53 replied:
Seems implementation is copied from torch/csrc/jit/register_special_ops.cpp?
What's our long term plan to support those ops in lite interpreter? Are we going to port all those ops to mobile?

iseeyuan replied:
The difference between jit/register_special_ops.cpp (or register_prim_ops.cpp) and here is that in jit, there is reference to node. I think they can only be merged when jit is also using OPN + vararg ops. It may be feasible when the type specific is removed. @zdevito, correct me if I'm wrong.

zdevito replied:
We should take care not to copy/paste operator implementations. In this care, for instance, we should have a single implementation of `format` and have the real and mobile interpreter call it.

zdevito replied:
Similar for the above -- we should not start copy/pasting implementations of operators just because they are OPN instructions. Instead, we should put these `void op(int num_inputs, Stack& stack)` functions into a compilation unit and have both the mobile and regular interpreter call them.

iseeyuan replied:
I'll make the essential functionalities shared between mobile and regular interpreter. I think that also address @ljk53 's concern.

Commit:
Update on "OPN ops TupleConstruct/Unpack and format."


TupleConstruct/Unpack as OPN ops.

zdevito reviewed (approved):
Looks good, but let's avoid copy/paste of implementations.

Commit:
Update on "OPN ops TupleConstruct/Unpack and format."


TupleConstruct/Unpack as OPN ops.

iseeyuan committed:
Update on "OPN ops TupleConstruct/Unpack and format."


TupleConstruct/Unpack as OPN ops.

Differential Revision: [D18499602](https://our.internmc.facebook.com/intern/diff/D18499602)

[ghstack-poisoned]

iseeyuan committed:
Update on "OPN ops TupleConstruct/Unpack and format."


TupleConstruct/Unpack as OPN ops.

Differential Revision: [D18499602](https://our.internmc.facebook.com/intern/diff/D18499602)

[ghstack-poisoned]

iseeyuan committed:
Update on "OPN ops TupleConstruct/Unpack and format."


TupleConstruct/Unpack as OPN ops.

Differential Revision: [D18499602](https://our.internmc.facebook.com/intern/diff/D18499602)

[ghstack-poisoned]

iseeyuan committed:
Update on "OPN ops TupleConstruct/Unpack and format."


TupleConstruct/Unpack as OPN ops.

Differential Revision: [D18499602](https://our.internmc.facebook.com/intern/diff/D18499602)

[ghstack-poisoned]

iseeyuan committed:
Update on "OPN ops TupleConstruct/Unpack and format."


TupleConstruct/Unpack as OPN ops.

Differential Revision: [D18499602](https://our.internmc.facebook.com/intern/diff/D18499602)

[ghstack-poisoned]
